File: packages/sdk-ts/src/viz/ascii.ts
/**
 * ASCII tree rendering — standalone function operating on ManifestInstance.
 *
 * Extracted from ManifestInstance to keep the kernel class focused.
 */

import type { ManifestInstance } from "../kernel/instance.js";

// ---------------------------------------------------------------------------
// ASCII Tree
// ---------------------------------------------------------------------------

export function asciiTree(mi: ManifestInstance): string {
  const tree = mi.dependencyTree();
  if (Object.keys(tree).length === 0) {
    return "(no dependencies)";
  }

  const lines: string[] = [`\u{1F4E6} ${mi.scope}`];
  const agentNames = Object.keys(tree).sort();

  for (let i = 0; i < agentNames.length; i++) {
    const agentName = agentNames[i];
    const info = tree[agentName] as Record<string, unknown>;
    const isLastAgent = i === agentNames.length - 1;
    const prefix = isLastAgent ? "\u2514\u2500\u2500 " : "\u251C\u2500\u2500 ";
    const childPrefix = isLastAgent ? "    " : "\u2502   ";

    lines.push(`${prefix}\u{1F916} ${agentName}`);

    const dependsOn = (info.depends_on ?? {}) as Record<string, Record<string, Record<string, unknown>>>;
    const depGroups = Object.entries(dependsOn);

    for (let j = 0; j < depGroups.length; j++) {
      const [depType, deps] = depGroups[j];
      const isLastGroup = j === depGroups.length - 1;
      const groupPrefix = isLastGroup ? "\u2514\u2500\u2500 " : "\u251C\u2500\u2500 ";
      const itemPrefix = isLastGroup ? "    " : "\u2502   ";

      lines.push(`${childPrefix}${groupPrefix}${depType}/`);

      const depItems = Object.entries(deps);
      for (let k = 0; k < depItems.length; k++) {
        const [depName, depInfo] = depItems[k];
        const isLastDep = k === depItems.length - 1;
        const depPrefix = isLastDep ? "\u2514\u2500\u2500 " : "\u251C\u2500\u2500 ";
        const kind = (depInfo.kind as string) ?? "";
        const found = (depInfo.found as boolean) ?? true;

        // Use KindPort.asciiIcon + graphMeta for kind-specific labels
        const depKp = mi.kindFor(kind);
        const depIcon = depKp?.asciiIcon ?? "";
        const depDoc = mi._one(kind, depName);
        const depMeta = depDoc ? (depKp?.graphMeta?.(depDoc) ?? {}) : {};
        let label: string;
        if ((depMeta as Record<string, unknown>).severity) {
          const m = depMeta as Record<string, unknown>;
          const severityIcon = m.severity === "error" ? "\u{1F534}" : "\u{1F7E1}";
          label = `${severityIcon} ${depName} (${m.rules ?? 0} rules, ${m.severity})`;
        } else {
          label = depIcon ? `${depIcon} ${depName}` : depName;
        }

        if (!found) {
          label += " \u274C MISSING";
        }

        lines.push(`${childPrefix}${itemPrefix}${depPrefix}${label}`);
      }
    }
  }

  return lines.join("\n");
}

File: packages/sdk-py/dna/viz/ascii.py
"""ASCII tree rendering — standalone function operating on ManifestInstance.

Extracted from ManifestInstance to keep the kernel class focused.
1:1 parity with TypeScript viz/ascii.ts.
"""
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dna.kernel.instance import ManifestInstance


def ascii_tree(mi: ManifestInstance) -> str:
    """Terminal-friendly dependency tree. No renderer needed."""
    tree = mi.dependency_tree()
    if not tree:
        return "(no dependencies)"

    lines = [f"\U0001f4e6 {mi.scope}"]

    agent_names = sorted(tree.keys())
    for i, agent_name in enumerate(agent_names):
        info = tree[agent_name]
        is_last_agent = i == len(agent_names) - 1
        prefix = "\u2514\u2500\u2500 " if is_last_agent else "\u251c\u2500\u2500 "
        child_prefix = "    " if is_last_agent else "\u2502   "

        lines.append(f"{prefix}\U0001f916 {agent_name}")

        dep_groups = list(info.get("depends_on", {}).items())
        for j, (dep_type, deps) in enumerate(dep_groups):
            is_last_group = j == len(dep_groups) - 1
            group_prefix = "\u2514\u2500\u2500 " if is_last_group else "\u251c\u2500\u2500 "
            item_prefix = "    " if is_last_group else "\u2502   "

            lines.append(f"{child_prefix}{group_prefix}{dep_type}/")

            dep_items = list(deps.items())
            for k, (dep_name, dep_info) in enumerate(dep_items):
                is_last_dep = k == len(dep_items) - 1
                dep_prefix = "\u2514\u2500\u2500 " if is_last_dep else "\u251c\u2500\u2500 "
                kind = dep_info.get("kind", "")
                found = dep_info.get("found", True)

                # Use KindPort.ascii_icon + graph_meta for kind-specific labels
                dep_kp = mi.kind_for(kind)
                dep_icon = getattr(dep_kp, "ascii_icon", "") or ""
                dep_doc = mi._kernel.get_document_sync(mi.scope, kind, dep_name)
                dep_meta = {}
                # KindPresentation.graph_meta — optional capability member,
                # typed access with default; summary is a core KindPort
                # member but dep_kp may be None (unknown kind name).
                graph_meta_fn = getattr(dep_kp, "graph_meta", None)
                summary_fn = getattr(dep_kp, "summary", None)
                if dep_doc and callable(graph_meta_fn):
                    dep_meta = graph_meta_fn(dep_doc) or {}
                elif dep_doc and callable(summary_fn):
                    dep_meta = summary_fn(dep_doc) or {}

                if dep_meta.get("severity"):
                    sev_icon = "\U0001f534" if dep_meta["severity"] == "error" else "\U0001f7e1"
                    rules_count = dep_meta.get("rules", 0)
                    label = f"{sev_icon} {dep_name} ({rules_count} rules, {dep_meta['severity']})"
                elif dep_icon:
                    label = f"{dep_icon} {dep_name}"
                else:
                    label = dep_name

                if not found:
                    label += " \u274c MISSING"

                lines.append(f"{child_prefix}{item_prefix}{dep_prefix}{label}")

    return "\n".join(lines)
